plugins/postcss-minify/src/index.ts

```diff
@@ -75,86 +85,107 @@ function removeEmptyNodes(node: Container | Document): boolean {
 	return false;
 }
 
+function setSemicolon(node: AtRule | Rule) {
+	if (node.raws.semicolon) {
+		const last = node.last;
+		if (last?.type !== 'decl' || !last.variable) {
+			node.raws.semicolon = false;
+		}
+	}
+}
+
+/** postcss-minify plugin options */
+export type pluginOptions = never;
+
 const creator: PluginCreator<pluginOptions> = () => {
 	return {
 		postcssPlugin: 'postcss-minify',
-		OnceExit(css) {
-			css.raws = {
-				before: '',
-				after: '\n',
-			};
+		prepare() {
+			const cache = new Map<string, string>();
+
+			return {
+				OnceExit(css) {
+					css.raws = {
+						before: '',
+						after: '\n',
+					};
 
-			css.walk((node) => {
-				if (node.type === 'comment') {
-					if (HAS_LEGAL_KEYWORDS.test(node.text) || HAS_SOURCE_MAP.test(node.text)) {
-						return;
-					}
+					css.walk((node) => {
+						if (node.type === 'comment') {
+							if (HAS_LEGAL_KEYWORDS.test(node.text) || HAS_SOURCE_MAP.test(node.text)) {
+								return;
+							}
 
-					node.remove();
-					return;
-				}
+							node.remove();
+							return;
+						}
 
-				if (node.type === 'decl' && node.variable) {
-					node.raws = {
-						...node.raws,
-						before: minify(node.raws?.before),
-					};
+						if (node.type === 'decl') {
+							if (node.variable) {
+								node.raws.before = '';
 
-					// never minify or modify variables
-					// browsers and CSSOM in particular handle these differently
-					return;
-				}
+								// never minify or modify variables
+								// browsers and CSSOM in particular handle these differently
+								return;
+							}
+						}
 
-				if (node.type === 'atrule') {
+						if (node.type === 'atrule') {
 
-					if (removeEmptyNodes(node)) {
-						return;
-					}
+							if (removeEmptyNodes(node)) {
+								return;
+							}
 
-					node.raws = {
-						before: minify(node.raws?.before),
-						after: minify(node.raws?.after),
-						between: minify(node.raws?.between),
-						semicolons: node.raws?.semicolons,
-						afterName: minify(node.raws?.afterName),
-					};
+							node.raws.after = '';
+							node.raws.afterName = ' ';
+							node.raws.before = '';
+							node.raws.between = '';
+							node.raws.params = undefined;
 
-					node.params = minify(node.params)!;
+							setSemicolon(node);
 
-					return;
-				}
+							node.params = minify(cache, node.params)!;
 
-				if (node.type === 'rule') {
+							return;
+						}
 
-					if (removeEmptyNodes(node)) {
-						return;
-					}
+						if (node.type === 'rule') {
 
-					node.raws = {
-						before: minify(node.raws?.before),
-						after: minify(node.raws?.after),
-						between: minify(node.raws?.between),
-						semicolons: node.raws?.semicolons,
-					};
```

Comment:
Assigning a completely new object to `raws` wasn't ideal.
- more memory allocations
- erasing unknown future fields

Especially the second part would have backfired. If PostCSS added a `foo` field in `raws` tomorrow we would always erase it and we would need to ship a new release.